src/Lecture4.hs

```diff
@@ -134,8 +140,22 @@ errors. We will simply return an optional result here.
 🕯 HINT: Use the 'readMaybe' function from the 'Text.Read' module.
 -}
 
+splitBy :: Char -> String -> [String]
+splitBy = go []
+  where
+    go :: String -> Char -> String -> [String]
+    go current _ [] = [current]
+    go current sep (x : xs)
+      | x == sep = current : splitBy sep xs
+      | otherwise = go (current ++ [x]) sep xs
+
 parseRow :: String -> Maybe Row
-parseRow = error "TODO"
+parseRow s = case parts of
+  ("" : _) -> Nothing
+  [a, b, c] -> (readMaybe b :: Maybe TradeType) >>= (\tt -> (readMaybe c :: Maybe Int) >>= (\cost -> if cost < 0 then Nothing else Just (Row a tt cost)))
+  _ -> Nothing
+  where
+    parts = splitBy ',' s
```

Comment:
You have an excellent monadic implementation 🔝 

I haven't explained the following during lectures but there's a nice way to write this parsing syntax using `do`-notation, `Monad` instance for `Maybe`, and the `MonadFail` desugaring trick 🧙🏻 

You can read more about this trick here:

* https://kowainik.github.io/posts/haskell-mini-patterns#monadfail-sugar

Applying this programming pattern and an existing helper function `guard`, the `parseRow` function will look like this:

```suggestion
parseRow s = do
    [a, b, c] <- Just $ splitBy ',' s
    guard $ a /= ""
    tt <- readMaybe b
    cost <- readMaybe c
    guard $ cost >= 0
    pure $ Row a tt cost
```